pkg/proxy/nftables/proxier.go

```diff
@@ -1222,36 +1212,67 @@ func (proxier *Proxier) syncProxyRules() {
 			// need the following rules to allow requests from this node.
 			if allowFromNode {
 				for _, lbip := range svcInfo.LoadBalancerVIPStrings() {
-					proxier.natRules.Write(
-						"-A", string(fwChain),
-						"-m", "comment", "--comment", fmt.Sprintf(`"%s loadbalancer IP"`, svcPortNameString),
-						"-s", lbip,
-						"-j", string(externalTrafficChain),
-					)
+					tx.Add(&knftables.Rule{
+						Chain: fwChain,
+						Rule: knftables.Concat(
+							ipX, "saddr", lbip,
+							"goto", externalTrafficChain,
+						),
+					})
 				}
 			}
 			// If the packet was able to reach the end of firewall chain,
 			// then it did not get DNATed, so it will match the
 			// corresponding KUBE-PROXY-FIREWALL rule.
-			proxier.natRules.Write(
-				"-A", string(fwChain),
-				"-m", "comment", "--comment", fmt.Sprintf(`"other traffic to %s will be dropped by KUBE-PROXY-FIREWALL"`, svcPortNameString),
-			)
+			tx.Add(&knftables.Rule{
+				Chain:   fwChain,
+				Rule:    "continue",
+				Comment: ptr.To("other traffic will be dropped by firewall"),
+			})
+		}
+
+		if svcInfo.SessionAffinityType() == v1.ServiceAffinityClientIP {
+			// Generate the per-endpoint affinity sets
+			for _, ep := range allLocallyReachableEndpoints {
+				epInfo, ok := ep.(*endpointInfo)
+				if !ok {
+					klog.ErrorS(nil, "Failed to cast endpointsInfo", "endpointsInfo", ep)
+					continue
+				}
+
+				// Create a set to store current affinity mappings. The
```

Comment:
Can we document clearly whether this is interpreting affinity in the iptables way (per-service, per-port) or ipvs way  (per-service) ?

Reply:
Sure. (It's the iptables way.)